### DISCUSSION TITLE
Is there a way to delete xcom in Airflow 2?

### DISCUSSION THREAD

**Discussion by marti1125** [apache/airflow#23793] [Q&A] [answered]
I am trying to delete xcom after Dag was finished, I didn't find how to delete xcoms for return_value

**Comment by kolfild26** [comment, marked as answer]
This will write the result of `BashOperator` to Xcom:
```
with DAG(dag_id="del_xcom",
         schedule_interval=None,
         start_date=days_ago(1)) as dag:
	
    bash = BashOperator(
    task_id='bash',
    bash_command="hostname -i | awk '{print $1}'",
    do_xcom_push=True)
    
    end = DummyOperator(task_id="end")
    
    bash >> end
```
And this will cleanup in the next task:
```
from airflow.models import XCom
from airflow.utils.db import provide_session

with DAG(dag_id="del_xcom",
         schedule_interval=None,
         start_date=days_ago(2)) as dag:

    @provide_session
    def clean_xcom(session=None, **context):
        dag = context["dag"]
        dag_id = dag._dag_id 
        session.query(XCom).filter(XCom.dag_id == dag_id).delete()

    delete_xcom = PythonOperator(
        task_id="delete_xcom",
        python_callable = clean_xcom, 
        dag=dag
    )
	
    bash = BashOperator(
    task_id='bash',
    bash_command="hostname -i | awk '{print $1}'",
    do_xcom_push=True)
    
    end = DummyOperator(task_id="end")
    
    bash >> delete_xcom >> end
```

**Reply by potiuk** [reply]
We don't have an official way to delete XCOM. It is there so for example you see it in the UI when you se details of the task. You could potentially write a piece of code to delete it but it is not guaranteed you will be able to run it as part of the task. You can look it up in stack overflow.

**Reply by potiuk** [reply]
I am confused. I am not sure what you are asking for. Could you be more specific what your question is ?

**Reply by josh-fell** [reply]
You could set `do_xcom_push=False` within the `default_args` of your DAGs or use a [Cluster Policy](https://airflow.apache.org/docs/apache-airflow/stable/concepts/cluster-policies.html) to set `do_xcom_push=False` for every task at load time. Both would alleviate the need to set `do_xcom_push` on _every_ task in your environment.

For example:
```python
# airflow_local_settings.py

from airflow.models import BaseOperator

def task_policy(task: BaseOperator) -> None:
    task.do_xcom_push = False
```

**Reply by chnaveen138** [reply]
Hi, I'm not sure if this is a bug in airflow, but I'm experiencing the below weird issue while using the code given above for clearing xcoms.

When I raise some exception immediately after `session.query(XCom).filter(...).delete()`, it is not deleting the XComs. If I remove exception line, then it is working as expected and clearing XComs correctly. What could be the reason for this? How to fix this? Any help would be appreciated.

```
from airflow.models import XCom
from airflow.utils.db import provide_session

with DAG(dag_id="del_xcom",
         schedule_interval=None,
         start_date=days_ago(2)) as dag:

    @provide_session
    def clean_xcom(session=None, **context):
        dag = context["dag"]
        dag_id = dag._dag_id 
        session.query(XCom).filter(XCom.dag_id == dag_id).delete()
        raise Exception("Some dummy error")

    delete_xcom = PythonOperator(
        task_id="delete_xcom",
        python_callable = clean_xcom, 
        dag=dag
    )
    
    end = DummyOperator(task_id="end")
    
    delete_xcom >> end
```

**Reply by Taragolis** [reply]
Manual run DB queries against Airflow Models like `session.query(XCom)...` it is not a part of the [Public Interface of Airflow ](https://airflow.apache.org/docs/apache-airflow/stable/public-airflow-interface.html).

Please consider to use cli command [`airflow db clean`](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#clean)

**Reply by chnaveen138** [reply]
Hi, I agree but in my use-case I need to clear xcoms of a task of specific dag_run as part of a Python operator. This is required for one of the functionalities and the above answer is suiting correctly for our use-case.
 
If I have to use CLI command, I might need to use separate operator like bash operator and do it. Is there any way to do that as part of same Python operator using cli command?

**Reply by Taragolis** [reply]
That is not implemented (or maybe i blind), so i can't see any way to do that in the backward compatibility method, the reason for that might be because it is not popular option and no one want to invest time to implements this yet.

Apache Airflow it is OSS software, so anyone who want to add some feature could do that https://github.com/apache/airflow/blob/main/CONTRIBUTORS_QUICK_START.rst

**Reply by Taragolis** [reply]
Another option [create feature request](https://github.com/apache/airflow/issues/new/choose) but make sure same request not exists, and maybe someone could implement these feature.

**Reply by chnaveen138** [reply]
Ok, will once check and raise a feature request. But I think I found the solution for the above problem.

After `session.query(XCom)...`, by keeping `session.commit()` line, it is clearing xcoms even after raising exception immediately after `session.query`. So I will be using this for now till that feature is available.

Here is updated code:
```
from airflow.models import XCom
from airflow.utils.db import provide_session

with DAG(dag_id="del_xcom",
         schedule_interval=None,
         start_date=days_ago(2)) as dag:

    @provide_session
    def clean_xcom(session=None, **context):
        dag = context["dag"]
        dag_id = dag._dag_id 
        session.query(XCom).filter(XCom.dag_id == dag_id).delete()
        # HAD TO ADD THIS LINE AFTER QUERY
        session.commit()
        raise Exception("Some dummy error")

    delete_xcom = PythonOperator(
        task_id="delete_xcom",
        python_callable = clean_xcom, 
        dag=dag
    )
    
    end = DummyOperator(task_id="end")
    
    delete_xcom >> end

**Reply by Taragolis** [reply]
Just be aware that [AIP-44](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-44+Airflow+Internal+API) would disable direct communications with DB

**Reply by chnaveen138** [reply]
Hmm, Noted. Thank you @Taragolis for mentioning this info and heads up.
